File: app/src/main/java/co/za/takeabyte/gradalot2020/productdescription/viewmodel/ViewModelProductDescription.java
package co.za.takeabyte.gradalot2020.productdescription.viewmodel;

import android.content.Context;
import android.os.Bundle;

import androidx.annotation.DrawableRes;
import androidx.annotation.Nullable;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.OnLifecycleEvent;
import androidx.lifecycle.ViewModel;

import co.za.takeabyte.gradalot2020.productdescription.IProductDescriptionView;
import co.za.takeabyte.gradalot2020.productdescription.ProductDescriptionActivity;

public class ViewModelProductDescription extends ViewModel implements LifecycleObserver {

    private Context context;

    private String title;

    @DrawableRes
    private int imageRes = -1;

    public void setContext(Context context) {
        this.context = context;
    }

    public void setDataBundle(@Nullable Bundle dataBundle) {
        if (dataBundle != null) {
            title = dataBundle.getString(ProductDescriptionActivity.PRODUCT_DESCRIPTION_TITLE);
            imageRes = dataBundle.getInt(ProductDescriptionActivity.PRODUCT_DESCRIPTION_IMAGE);
        }
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    public void onStart() {
        renderUI();
    }

    private void renderUI() {
        if (context instanceof IProductDescriptionView) {
            IProductDescriptionView productDescriptionView = (IProductDescriptionView) context;

            if (imageRes == -1) {
                //render placeholder
            } else {
                productDescriptionView.renderImage(imageRes);
            }

            productDescriptionView.renderTitle(title);
        }
    }

}

File: app/src/main/java/co/za/takeabyte/gradalot2020/productdescription/viewmodel/ViewModelProductDescription.kt
package co.za.takeabyte.gradalot2020.productdescription.viewmodel

import android.content.Context
import android.os.Bundle
import androidx.annotation.DrawableRes
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.OnLifecycleEvent
import androidx.lifecycle.ViewModel
import co.za.takeabyte.gradalot2020.productdescription.IProductDescriptionView
import co.za.takeabyte.gradalot2020.productdescription.ProductDescriptionActivity

class ViewModelProductDescription : ViewModel(), LifecycleObserver {

    private var context: Context? = null

    private var title: String = String()

    @DrawableRes
    private var imageUrl: Int = -1

    fun setContext(context: Context) {
        this.context = context
    }

    fun setDataBundle(data: Bundle?) {
        title = data?.getString(ProductDescriptionActivity.PRODUCT_DESCRIPTION_TITLE) ?: ""
        imageUrl = data?.getInt(ProductDescriptionActivity.PRODUCT_DESCRIPTION_IMAGE) ?: -1
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    fun onStart() {
        renderUI()
    }

    private fun renderUI() {
        val context = context
        if (context is IProductDescriptionView) {
            if (imageUrl == -1) {
//                render placeholder
            } else {
                context.renderImage(imageUrl)
            }

            context.renderTitle(title)
        }
    }

}